### PR TITLE
Add univ3 router api

### PR DESCRIPTION
**Pull request by vkgnosis**
Api for https://github.com/cowprotocol/univ3-api which will be used as a solver.

### Test Plan

CI

### CODE REVIEW

**Review by MartinquaXD — APPROVED** [review]
Cool to have this uniswap integration!

File: crates/shared/src/univ3_router_api.rs [hunk 1]
```diff
@@ -0,0 +1,115 @@
+use crate::solver_utils::deserialize_prefixed_hex;
+use anyhow::{Context, Result};
+use model::u256_decimal;
+use primitive_types::{H160, U256};
+use reqwest::{Client, Url};
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Copy, Clone, Serialize)]
```
Comment (sunce86) [inline crates/shared/src/univ3_router_api.rs:8]: nit: `#[serde(rename_all = "camelCase")]`

**Review by sunce86 — APPROVED** [review]

File: crates/shared/src/univ3_router_api.rs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+pub struct Api {
+    client: Client,
+    estimate: Url,
+}
+
+impl Api {
+    pub fn new(client: Client, base: Url) -> Self {
+        Self {
+            client,
+            estimate: base.join("estimate").unwrap(),
```
[the remark this answers is not in the document]
Comment (vkgnosis) [reply crates/shared/src/univ3_router_api.rs:46]: I documented that the endpoint is /estimate in the router's readme. Technically it replies this way to all requests but I want to have the option to add other routes in the future.

File: crates/shared/src/univ3_router_api.rs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        let response: Response = serde_json::from_str(json).unwrap();
+        assert_eq!(response.quote, 284671676368727715u64.into());
+        assert_eq!(response.gas, 113000);
+        assert_eq!(response.call_data, &[0x01, 0x02]);
+    }
+
+    #[tokio::test]
+    #[ignore]
+    async fn real() {
+        let api = Api::new(Default::default(), "http://localhost:8080".parse().unwrap());
```
[the remark this answers is not in the document]
Comment (vkgnosis) [reply crates/shared/src/univ3_router_api.rs:104]: I don't want to change the default port of other standalone projects to match this one. When you're running the test here you're probably not running the orderbook right now anyway.

File: crates/shared/src/univ3_router_api.rs [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct Response {
+    #[serde(with = "u256_decimal")]
+    pub quote: U256,
+    #[serde(with = "serde_with::rust::display_fromstr")]
+    pub gas: u64,
+    #[serde(deserialize_with = "deserialize_prefixed_hex")]
+    pub call_data: Vec<u8>,
```
[the remark this answers is not in the document]
Comment (vkgnosis) [reply crates/shared/src/univ3_router_api.rs:34]: I disagree with the use of Bytes but haven't consistently mentioned it in reviews. Bytes is is a "random" type from the web3 crate. No where do they document what it is used for and how it is serialized. So we shouldn't be relying on that. To me Vec<u8> is already bytes and how it is serialized is what you specify with the serde tag just like we do with the u64 above.

File: crates/shared/src/univ3_router_api.rs (hunk 1 above)
Comment (vkgnosis) [reply crates/shared/src/univ3_router_api.rs:8]: For this one I felt giving the enum values explicitly was a bit easier to understand.

**Pull request merged**